hail/python/hailtop/auth/auth.py

```diff
@@ -1,74 +1,143 @@
+from hailtop.aiocloud.common.credentials import AnonymousCloudCredentials, CloudCredentials, Credentials
 from hailtop.aiocloud.common import Session
-from hailtop.config import get_deploy_config, DeployConfig
+from hailtop.aiocloud.aiogoogle import GoogleCredentials
+from hailtop.aiocloud.aioazure import AzureCredentials
+from hailtop.config import get_deploy_config, DeployConfig, get_user_identity_config_path
 from hailtop.utils import async_to_blocking, retry_transient_errors
 
-from .tokens import Tokens, get_tokens
+from .tokens import get_tokens, Tokens
 
 
-class HailStoredTokenCredentials(CloudCredentials):
-    def __init__(self, tokens: Tokens, namespace: Optional[str], authorize_target: bool):
-        self._tokens = tokens
-        self._namespace = namespace
-        self._authorize_target = authorize_target
+class IdentityProvider(Enum):
+    GOOGLE = 'Google'
+    MICROSOFT = 'Microsoft'
 
-    @staticmethod
-    def from_file(credentials_file: str, *, namespace: Optional[str] = None, authorize_target: bool = True):
-        return HailStoredTokenCredentials(get_tokens(credentials_file), namespace, authorize_target)
+
+@dataclass
+class IdentityProviderSpec:
+    idp: IdentityProvider
+    # Absence of specific oauth credentials means Hail should use latent credentials
+    oauth2_credentials: Optional[dict]
 
     @staticmethod
-    def default_credentials(*, namespace: Optional[str] = None, authorize_target: bool = True):
-        return HailStoredTokenCredentials(get_tokens(), namespace, authorize_target)
+    def from_json(config: Dict[str, Any]):
+        return IdentityProviderSpec(IdentityProvider(config['idp']), config.get('credentials'))
+
+
+class HailCredentials(Credentials):
+    def __init__(self, tokens: Tokens, cloud_credentials: Optional[CloudCredentials], namespace: str, authorize_target: bool):
+        self._tokens = tokens
+        self._cloud_credentials = cloud_credentials
+        self._namespace = namespace
+        self._authorize_target = authorize_target
 
     async def auth_headers(self) -> Dict[str, str]:
-        deploy_config = get_deploy_config()
-        ns = self._namespace or deploy_config.default_namespace()
-        return namespace_auth_headers(deploy_config, ns, self._tokens, authorize_target=self._authorize_target)
+        headers = {}
+        if self._authorize_target:
+            token = await self._get_idp_access_token_or_hail_token(self._namespace)
+            headers['Authorization'] = f'Bearer {token}'
+        if get_deploy_config().location() == 'external' and self._namespace != 'default':
+            # We prefer an extant hail token to an access token for the internal auth token
+            # during development of the idp access token feature because the production auth
+            # is not yet configured to accept access tokens. This can be changed to always prefer
+            # an idp access token when this change is in production.
```

Comment:
Can you create an issue to fully switch over to access tokens? I want to make sure we don't forget to complete the move.

Reply:
Made #13531